Convert "Create a DevTools extension" to sample-based (#3616)
MicrosoftDocs/edge-developer: microsoft-edge/extensions/developer-guide/devtools-extension.md

Reviewer: As discussed, let's merge this with **Code for Custom DevTools tool**. This article seems like an unnecessary duplicate.

| `icon.png` | Icon to display on the tool's tab in the Activity bar of DevTools and in the **More tools** menu. |
| `README.md` | Basic information for developers about how to use the sample. |

<!-- keep rows together: devtools.html, then devtools.js -->

Details are below.

<!-- ====================================================================== -->
## Interact between the webpage and DevTools

Multiple files are needed to enable interaction between the inspected webpage and DevTools.  These files run in different contexts:

* `devtools.js` runs in the context of the custom tool panel, in DevTools.

* `content_script.js` is a _content script_, which means that it runs in the context of the inspected webpage.  In the same way that other scripts are loaded by the webpage, a content script has access to the DOM and can change it.

The DevTools page, inspected page, and content script fit together in an extension:

> ![Diagram showing the anatomy of a DevTools extension](./devtools-extension-images/architecture.png)<!-- todo: omit SW -->

Author: omit Service Worker from diagram

https://review.learn.microsoft.com/en-us/microsoft-edge/extensions/developer-guide/devtools-extension?branch=pr-en-us-3616#interact-between-the-webpage-and-devtools

direct png link:
https://review.learn.microsoft.com/en-us/microsoft-edge/extensions/developer-guide/devtools-extension-images/architecture.png?branch=pr-en-us-3616

@captainbrosset how to proceed?

Author: removed SW from diagram

author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: article
ms.service: microsoft-edge
ms.subservice: extensions
ms.date: 04/20/2026
---
# Add a custom tool in Microsoft Edge DevTools
<!-- https://learn.microsoft.com/microsoft-edge/extensions/developer-guide/devtools-extension -->
<!-- upstream equiv: https://developer.chrome.com/docs/extensions/how-to/devtools/extend-devtools#devtools_extension_examples -->

This article describes the code of the Custom DevTools sample.  To first install and run the sample, see [Sample: Custom DevTools tool](./devtools-extension-sample.md).

<!-- ====================================================================== -->
## Introduction

Like a regular extension, a DevTools extension has a manifest file.  That's described in [Get started developing an extension](../getting-started/index.md), which presents a basic extension sample that you can start with.

An extension that extends DevTools additionally includes:
* A rendered HTML file to display in the tool's panel in DevTools.

> * A non-rendered HTML file that's run when DevTools is opened, that just loads a DevTools .js file.

Reviewer: ```suggestion
* A non-rendered HTML file that runs when DevTools is opened, and which typically handles communication between the custom DevTools tool and the inspected webpage.
```

> The present article explains the code in [Sample: Custom DevTools tool](./devtools-extension-sample.md).

Reviewer: This is almost a dup of line 15. Maybe remove this line.

Author: removed line

> To add a custom tool in Microsoft Edge DevTools, create a Microsoft Edge extension that adds a custom tool in Microsoft Edge DevTools, including a tool tab and panel.

Reviewer: This feels circular: "To add a tool in DevTools, create an extension that adds a tool in DevTools".
Maybe remove this sentence.

Author: removed.  contains a substantive point, but other content amounts to the same point

![Custom tab](./devtools-extension-images/custom-tab-activity-bar.png)

A _panel_ is a tool page in Microsoft Edge DevTools, along with the tool's tab in the **Activity Bar**.

See also:
* [Sample: Picture viewer pop-up webpage](../samples/picture-viewer-popup-webpage.md)
* [Overview of DevTools](../../devtools/overview.md)<!-- long jump -->

<!-- ====================================================================== -->
## Overview of files

The DevTools extension sample has the following files:

| File | Description |
|---|---|
| `manifest.json` | Information about the extension: name, description, version, manifest version, and HTML page to show in DevTools. |
| `panel.html` | Webpage to display in the custom panel in DevTools. |
| `devtools.html` | A non-rendered HTML file that's run when DevTools is opened, to load the extension's JavaScript files. |
| `devtools.js` | Main logic for the custom DevTools page. |
| `service-worker.js` | A service worker that sets up event listeners for communications between the inspected page and DevTools. |

> | `content_script.js` | Code which the custom DevTools tool injects in the inspected webpage.  Prints a message<!-- todo: what message? --> to the console when the script is injected in the page.  Adds a click event listener to the page that will send a message with mouse-click position in the inspected page. |

Reviewer: ```suggestion
| `content_script.js` | Code which the extension injects in the inspected webpage.  Prints a message<!-- todo: what message? --> to the console when the script is injected in the page.  Adds a click event listener to the page that will send a message with mouse-click position in the inspected page. |
```

Details are below.

<!-- ====================================================================== -->
## Interact between the webpage and DevTools

Code<!-- todo: which file of sample? --> that interacts with the inspected webpage does the following:

1. Listens to click events that happen on the webpage and log them into the DevTools **Console** tool.

1. Displays the mouse-click coordinates in the Custom DevTools tool.

1. When the user clicks a button in the Custom DevTools tool, a greeting alert is displayed in the inspected webpage.

The Custom DevTools tool has direct access to the inspected webpage, and runs when DevTools is opened.  The sample uses a content script (`content_script.js`) and a background service worker (`service-worker.js`), as well as the main JS file, `devtools.js`:<!-- todo: why are the content script `content_script.js` and the service worker script and the main js file broken out this way? -->

* A _content script_ (such as `content_script.js`) runs in the context of the inspected webpage.  In the same way that other scripts are loaded by the webpage, a content script has access to the DOM and can change it.

> * A _background service worker_ (such as `service-worker.js`) is a script that the browser runs in a separate thread.  This script has access to the Microsoft Edge extension APIs.

Reviewer: Maybe reverse things and start with the list of files that are involved in the communication between the webpage and devtools.

Something like this:

```suggestion
Multiple files are needed to enable interaction between the inspected webpage and DevTools. These files run in different contexts:

* `devtools.js` runs in the context of the custom tool panel, in DevTools.

* `content_script.js` is a _content script_, which means that it runs in the context of the inspected webpage.  In the same way that other scripts are loaded by the webpage, a content script has access to the DOM and can change it.

* `service-worker.js` is a _background service worker_. It runs in a separate thread and has access to extension APIs.
```

<!-- ====================================================================== -->

> ## Architecture

Reviewer: With my rewrite proposal, above, you could merge this h2 section with the previous h2 section.

Author: merging, by deleting:
```
<!-- ====================================================================== -->
## Architecture
```

The DevTools page, inspected page, content script, and background service worker fit together in an extension:

![Diagram showing the anatomy of a DevTools extension](./devtools-extension-images/architecture.png)

The `content_script.js` detects where the user clicks on the inspected webpage:

```javascript
document.addEventListener("click", (event) => {
  chrome.runtime.sendMessage({
      click: true,
      xPosition: event.clientX + document.body.scrollLeft,

>       yPosition: event.clientY + document.body.scrollTop

Reviewer: ```suggestion
document.addEventListener("click", event => {
  chrome.runtime.sendMessage({
    click: true,
    xPosition: event.clientX + document.body.scrollLeft,
    yPosition: event.clientY + document.body.scrollTop
  });
});
```

    <h3>Send message from inspected page to DevTools</h3>
    <p>Click somewhere in the inspected webpage.</p>
    <div>
      <b>Coordinates:</b> <span id="youClickedOn"></span>
    </div>

  </body>
</html>
```

`panel.html` is referenced in the `chrome.devtools.panels.create` method call in `devtools.js`.  The `panel.html` webpage contains the user interface of the custom tool's panel.

The above elements demonstrate the interaction between the inspected page, the DevTools panel, and the background service worker.

When the user clicks the `sayHello` button in the DevTools extension, an alert is displayed in the inspected window.

When the user clicks anywhere in the inspected page, the mouse-click coordinates are displayed in the DevTools custom tool.

See also:
* [panel.html](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/panel.html) - source file in repo.

<!-- keep sections together: devtools.html, then devtools.js -->
<!-- ====================================================================== -->
## `devtools.html`

A non-rendered HTML file that's run when DevTools is opened, to load the extension's JavaScript files.

> This is a non-rendered HTML file that's run when DevTools is opened, to load the extension's `devtools.js` file.  Such a file is required.

Reviewer: ```suggestion
A non-rendered HTML file that's run when DevTools is opened.  This file is required.
```

> The file name `devtools.html` matches the `devtools_page` member in the manifest file:

Reviewer: ```suggestion
The file name `devtools.html` must match the value of the `devtools_page` member in the manifest file.
```

> A `<script>` element loads the main JavaScript file that's used directly by `panel.html`.  This `devtools.html` file isn't displayed in DevTools; only the `panel.html` file is displayed in DevTools.

Reviewer: ```suggestion
`devtools.html` contains a `<script>` element which loads `devtools.js`:
```

`devtools.html`:

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
  </head>
  <body>
    <script src="devtools.js"></script>
  </body>
</html>
```

> In the manifest file (`manifest.json`), the `devtools_page` field specifies the above file (`devtools.html`).  `devtools.html`, above, contains a `<script>` element that loads `devtools.js`.

Reviewer: ```suggestion
```

Remove. The same content is already mentioned above.

See also:
* [devtools.html](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/devtools.html) - source file in repo.

<!-- ====================================================================== -->
## `devtools.js`

`devtools.js` contains the main logic for the custom DevTools page.  `devtools.js` does the following:

> 1. Creates a new **Custom** tool tab and panel in DevTools.

Reviewer: ```suggestion
1. Creates a new **Custom** tool tab and panel in DevTools by using `chrome.devtools.panels.create()`.
```

    if (sender.tab && request.click == true) {
        console.log('I am here!');
        if (youClickedOn) {
            youClickedOn.innerHTML = `(${request.xPosition}, ${request.yPosition})`;
        }
        sendResponse({
            xPosition: request.xPosition,
            yPosition: request.yPosition
        });
    }
});

// Create a connection to the background service worker.
const backgroundPageConnection = chrome.runtime.connect({
    name: "devtools-page"
});

// Relay the tab ID to the background service worker.
backgroundPageConnection.postMessage({
    name: 'init',
    tabId: chrome.devtools.inspectedWindow.tabId
});
```

The `create` method call in `devtools.js`:

```javascript
chrome.devtools.panels.create("Custom", "icon.png", "panel.html", panel => {
    // code invoked on panel creation
});
```

The `create` method has the following signature:

```javascript
chrome.devtools.panels.create(
    title: string, // Tool tab's label in Activity bar.
    iconPath: string, // Icon to display in tool's tab.
    pagePath: string, // Webpage to display in tool's panel.
    callback: function // Code to run when tool is opened.
)
```

> `devtools.js` does the following:

Reviewer: The exact same sentence appears at line 265. Should we combine the 2 "devtools.js does this" lists?
Or maybe change this sentence to be more specific about what appears in the list:

```suggestion
To communicate with the inspected webpage, `devtools.js` does the following:
```

Author: used:
To communicate with the inspected webpage, `devtools.js` does the following:

1. Uses the `chrome.runtime.connect` method to create a connection to the background service worker.

1. Sends the inspected window `tabId` to the service worker by using the `backgroundPageConnection.postMessage` method.

1. Adds event listeners to the `sayHello` button and `youClickedOn` label that's defined in the `panel.html` file.

> When the user clicks the `sayHello` button, the DevTools extension runs a code snippet of `alert("Hello from the DevTools Extension");` in the inspected window by invoking the `eval()` method of the inspected window `chrome.devtools.inspectedWindow`.

Reviewer: ```suggestion
When the user clicks the `sayHello` button, `devtools.js` runs code directly in the inspected window, by using the `chrome.devtools.inspectedWindow.eval()` method.
```

When the user clicks anywhere in the inspected window, the DevTools extension will receive a message, from the background service worker, with `request.click == true` and the mouse position information.

See also: 
* [devtools.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/devtools.js) - source file in repo.
* [Supported APIs for Microsoft Edge extensions](../developer-guide/api-support.md)
   * [chrome.devtools.panels](https://developer.chrome.com/docs/extensions/reference/api/devtools/panels)
      * [create()](https://developer.chrome.com/docs/extensions/reference/api/devtools/panels#method-create)

<!-- ====================================================================== -->
## `service-worker.js`

A background service worker that sets up event listeners for communications between the inspected page and DevTools.

> A _background service worker_ is a script that the browser runs in a separate thread.  This script has access to the Microsoft Edge extension APIs.

Reviewer: ```suggestion
A _background service worker_ is a script that the browser runs in a separate thread.  This script has access to the extension APIs and sets up event listeners for communications between the inspected page and DevTools.
```

The DevTools Extension sample displays a custom tool in DevTools, including a **Custom** tab in the **Activity Bar** and a panel that's connected to the tab:

![Custom tab](./devtools-extension-images/custom-tab-activity-bar.png)

A _panel_ is a tool page in Microsoft Edge DevTools, along with the tool's tab in the **Activity Bar**.

See also:
* [Sample: Picture viewer pop-up webpage](../samples/picture-viewer-popup-webpage.md)
* [Overview of DevTools](../../devtools/overview.md)<!-- long jump -->

<!-- ====================================================================== -->
## Overview of files

The DevTools extension sample has the following files:

| File | Description |
|---|---|
| `manifest.json` | Information about the extension: name, description, version, manifest version, and HTML page to show in DevTools. |
| `panel.html` | Webpage to display in the custom panel in DevTools. |
| `devtools.html` | A non-rendered HTML file that's run when DevTools is opened, to load the extension's JavaScript files. |
| `devtools.js` | Main logic for the custom DevTools page. |

> | `service-worker.js` | A service worker that sets up event listeners for communications between the inspected page and DevTools. |

Reviewer: Remove this, see https://github.com/MicrosoftEdge/Demos/pull/106#discussion_r3310824276

> | `content_script.js` | Code which the extension injects in the inspected webpage.  Prints a message<!-- todo: what message? --> to the console when the script is injected in the page.  Adds a click event listener to the page that will send a message with mouse-click position in the inspected page. |

Author: What message?

Author: Nothing happens WHEN the script is injected.

when page is clicked, the message appears in console: "Received response"
`    response => {
      console.log("Received response", response);
    }
`
and also in custom tool, the coordinates of click are printed.

> | `content_script.js` | Code which the extension injects in the inspected webpage.  Prints a message<!-- todo: what message? --> to the console when the script is injected in the page.  Adds a click event listener to the page that will send a message with mouse-click position in the inspected page. |

Author: ```suggestion
| `content_script.js` | Code which the extension injects in the inspected webpage.  Adds a click event listener to the page that will send a message with the mouse-click position in the inspected page.  Sends a message to ...  |
```
sends a message, and devtools.js is listenging for that message as foolows:

```
chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
    // Messages from content scripts should have sender.tab set.
    if (sender.tab && request.click == true) {
        console.log('I am here!');
        if (youClickedOn) {
            youClickedOn.innerHTML = `(${request.xPosition}, ${request.yPosition})`;
        }
        sendResponse({
            xPosition: request.xPosition,
            yPosition: request.yPosition
        });
    }
});
```

Reviewer: ```suggestion
| `content_script.js` | Code which the extension injects in the inspected webpage. Adds a click event listener to the page that will send a message with the mouse-click position.  `devtools.js` listens to this message and displays the position in the custom panel. |
```

| `icon.png` | Icon to display on the tool's tab in the Activity bar of DevTools and in the **More tools** menu. |
| `README.md` | Basic information for developers about how to use the sample. |

<!-- keep rows together: devtools.html, then devtools.js -->

Details are below.

<!-- ====================================================================== -->
## Interact between the webpage and DevTools

Multiple files are needed to enable interaction between the inspected webpage and DevTools.  These files run in different contexts:

* `devtools.js` runs in the context of the custom tool panel, in DevTools.

* `content_script.js` is a _content script_, which means that it runs in the context of the inspected webpage.  In the same way that other scripts are loaded by the webpage, a content script has access to the DOM and can change it.

> * `service-worker.js` is a _background service worker_. It runs in a separate thread and has access to extension APIs.

Reviewer: We should remove since the service worker doesn't play any role in this extension. See https://github.com/MicrosoftEdge/Demos/pull/106#discussion_r3310824276

1. Sends the inspected window `tabId` to the service worker by using the `backgroundPageConnection.postMessage` method.

1. Adds event listeners to the `sayHello` button and `youClickedOn` label that's defined in the `panel.html` file.

When the user clicks the `sayHello` button, `devtools.js` runs code directly in the inspected window, by using the `chrome.devtools.inspectedWindow.eval()` method.

When the user clicks anywhere in the inspected window, the DevTools extension will receive a message, from the background service worker, with `request.click == true` and the mouse position information.

See also: 
* [devtools.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/devtools.js) - source file in repo.
* [Supported APIs for Microsoft Edge extensions](../developer-guide/api-support.md)
   * [chrome.devtools.panels](https://developer.chrome.com/docs/extensions/reference/api/devtools/panels)
      * [create()](https://developer.chrome.com/docs/extensions/reference/api/devtools/panels#method-create)

<!-- ====================================================================== -->

> ## `service-worker.js`

Reviewer: See https://github.com/MicrosoftEdge/Demos/pull/106#discussion_r3310824276
This file doesn't seem to actually be needed. So we might just want to remove this whole section.

        }
    };

    // Listen to messages sent from the DevTools page
    devToolsConnection.onMessage.addListener(devToolsListener);

    devToolsConnection.onDisconnect.addListener(() => {
        devToolsConnection.onMessage.removeListener(devToolsListener);
    });
});
```

The above code connects the background service worker with the DevTools page.  It listens to when the DevTools page connects, saves the connection, and sends a response back to the DevTools page.

This is useful when your background service worker is collecting data or performing tasks in the background that you want to be available in your DevTools extension.

See also:
* [service-worker.js](https://github.com/MicrosoftEdge/Demos/blob/main/devtools-extension/service-worker.js) - source file in repo.
* [Content scripts](https://developer.chrome.com/docs/extensions/develop/concepts/content-scripts) - Chrome Extensions docs.

<!-- ====================================================================== -->
## `content_script.js`

Code which the custom DevTools tool injects in the inspected webpage.
Contains JavaScript that's injected into the inspected webpage (any webpage).

This file does the following:

> * Prints a message<!-- todo: what message? --> to the DevTools **Console** when the page is clicked.

Author: Prints what message?

Author: seq:
1. msg is sent from content_script.js, received by devtools.js.
2. displays the coordinates in the custome tool in devtools, in panel.html
3. reply from devtools.js is received by content_script.js.
4. prints to console:       console.log("Received response", response); to the console

<img width="1920" height="1032" alt="image" src="https://github.com/user-attachments/assets/c85fa5c4-fe16-4298-9288-cd919240336f" />

Reviewer: Remove this step. Nothing is printed.

* Adds a click event listener to the inspected webpage, that sends a message containing the mouse-click position, by calling `chrome.runtime.sendMessage`.
   * Listens for the page click event via an event listener.

> * Adds a click event listener to the inspected webpage; clicking the page sends an event, caught by the event listener, which then sends a message to the WebView2 Runtime.

Reviewer: ```suggestion
* Adds a click event listener to the inspected webpage, that sends a message containing the mouse-click position, by calling `chrome.runtime.sendMessage`.
   * Listens for the page click event via an event listener.
```

That second bullet point was wrong, it's partly a duplicate of the first point, and mentions WebView2, which is wrong in this context.
